dns-secondary-operator/src/charm.py

```diff
@@ -0,0 +1,186 @@
+        self.framework.observe(self.on.collect_unit_status, self._on_collect_status)
+        self.framework.observe(
+            self.on["dns-transfer"].relation_joined, self._on_dns_transfer_relation_joined
+        )
+        self.framework.observe(
+            self.on["dns-transfer"].relation_changed, self._on_dns_transfer_relation_changed
+        )
+        self.unit.open_port("tcp", 53)  # Bind DNS
+        self.unit.open_port("udp", 53)  # Bind DNS
+
+    def _on_collect_status(self, event: ops.CollectStatusEvent) -> None:
+        """Handle collect status event.
+
+        Args:
+            event: Event triggering the collect-status hook
+        """
+        if not self._has_required_integration():
+            event.add_status(ops.BlockedStatus(STATUS_REQUIRED_INTEGRATION))
+        event.add_status(ops.ActiveStatus(""))
+
+    def _on_dns_transfer_relation_joined(self, _: ops.RelationJoinedEvent) -> None:
+        """Handle changed relation joined event."""
+        self._reconcile()
+
+    def _on_dns_transfer_relation_changed(self, _: ops.RelationChangedEvent) -> None:
+        """Handle changed relation changed event."""
+        self._reconcile()
+
+    def _on_config_changed(self, _: ops.ConfigChangedEvent) -> None:
+        """Handle changed configuration event."""
+        self._reconcile()
+
+    def _on_stop(self, _: ops.StopEvent) -> None:
+        """Handle stop."""
+        self.bind.stop()
+
+    def _reconcile(self) -> None:
+        """Reconcile the charm."""
+        if not self._has_required_integration():
+            return
+
+        try:
+            t = self.topology.current()
+        except topology.TopologyUnavailableError as err:
+            logger.info("Could not retrieve network topology: %s", err)
+            return
+        if self.unit.is_leader() and not t.is_current_unit_active:
+            self._check_and_may_become_active(t)
+
+        self.unit.status = ops.MaintenanceStatus("Preparing bind")
+        self.bind.setup(self.unit.name)
+        self.bind.start()
+
+        relation = self.model.get_relation(self.dns_transfer.relation_name)
+        data = self.dns_transfer.get_remote_relation_data()
+        self.bind.update_config_and_reload(data.zones, [str(a) for a in data.addresses])
+
+        if self.unit.is_leader():
+            requirer_addresses = str(self.config["ips"]).split(",")
+            requirer_data = dns_transfer.DNSTransferRequirerData(addresses=requirer_addresses)
+            self.dns_transfer.update_relation_data(relation, requirer_data)
+
+    def _has_required_integration(self) -> bool:
+        """Check if dns_transfer required integration is set.
+
+        Returns:
+            true if dns_transfer is set.
+        """
+        data = self.dns_transfer.get_remote_relation_data()
+        if data is None:
+            return False
+        return True
+
+    def _check_and_may_become_active(self, t: topology.Topology) -> bool:
```

Comment:
I don't think there's a need to keep the 'active'/'standby' dichotomy for this charm

Reply:
wouldnt that affect how topology is created?

Reply:
Secondaries also need to communicate their IP addresses back to the primary through `dns_transfer`. So topology can be useful in the reconcile loop